### PR TITLE
ci: Copy properties from issue to pull request

### LINKED ISSUE #1092: GitHub action to validate is any user is assigned to pull request
**Author the user story for this feature**

As a pull request author, I want to be informed when I forgot to assign my self to pull request so I can avoid empty author on release notes.

**Is your feature request related to a problem? Please describe.**

I forgot to assign myself to this pr: https://github.com/Flank/flank/pull/1087this cause no author assigned in release notes,

```

- [#1087](https://github.com/Flank/flank/pull/1087) Fix release action ()

```
https://github.com/Flank/flank/pull/1089

**Describe the solution you'd like**
Action should fail when no one is assigned to pull requests.

**Describe alternatives you've considered**

The alternative is assign the author of the first commit in pull request.

### PR DESCRIPTION
**Pull request by piotradamczyk5**
Fixes #1092

When creating pull request labels, assignees and estimates are copied from source issue.
Source issue is discovered by:
- Having reference in description (ex. Fixes #1092)
or
- Having reference in branch name(ex. #1092_copy_properties_to_pull_request)

To test it locally use command:
`flankScripts pullRequest copyProperties --github-token=<github token> --zenhub-token=<zenhub token, contact me if you would like to test it> --pr-number=<number of PR>
`
## Checklist

- [x] Find source issue by description or branch name
- [x] Copy Labels
- [x] Copy assignees
- [x] Copy estimations
- [x] Attach to github action
- [x] Documented
- [x] Unit tested

### CODE REVIEW

**Comment by bootstraponline** [comment]
> * Having reference in description (ex. Fixes #1092)

I think we always want a reference to the ticket in the description for pull requests related to an issue (this is how GitHub links PRs to issues). GitHub doesn't understand the branch convention.

**Comment by piotradamczyk5** [comment]
> > * Having reference in description (ex. Fixes #1092)
> 
> I think we always want a reference to the ticket in the description for pull requests related to an issue (this Yis how GitHub links PRs to issues). GitHub doesn't understand the branch convention.

Yes, the branch is just a fallback :) 99% of cases it will take it from description

File: flank-scripts/src/main/kotlin/flank/scripts/pullrequest/SetLabels.kt [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        .onError { println("Could not copy labels because of ${it.message}") }
+        .map { it.map { label -> label.name } }
+        .getOrNull()
+        ?.run { setLabelsToPullRequest(githubToken, pullRequestNumber, this) }
+}
+
+private suspend fun getLabelsFromIssue(githubToken: String, issueNumber: Int) =
+    Fuel.get("https://api.github.com/repos/Flank/flank/issues/$issueNumber/labels")
+        .appendHeader("Accept", "application/vnd.github.v3+json")
+        .appendHeader("Authorization", "token $githubToken")
```
Comment (adamfilipow92) [inline flank-scripts/src/main/kotlin/flank/scripts/pullrequest/SetLabels.kt:27]: In ```GitHubApi.kt``` we have function ```appendHeaders```. This function adds these headers to the request. Maybe we could use these functions in this and other places? WDYT?

Comment (piotradamczyk5) [reply flank-scripts/src/main/kotlin/flank/scripts/pullrequest/SetLabels.kt:27]: I think that this is very good idea

File: flank-scripts/src/main/kotlin/flank/scripts/pullrequest/SetAssignees.kt [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+import com.github.kittinunf.fuel.coroutines.awaitStringResult
+import com.github.kittinunf.result.onError
+import com.github.kittinunf.result.success
+import flank.scripts.utils.toJson
+import kotlinx.serialization.Serializable
+
+suspend fun setAssigneesToPullRequest(githubToken: String, pullRequestNumber: Int, assignees: List<String>) {
+    Fuel.post("https://api.github.com/repos/Flank/flank/issues/$pullRequestNumber/assignees")
+        .appendHeader("Accept", "application/vnd.github.v3+json")
+        .appendHeader("Authorization", "token $githubToken")
```
Comment (adamfilipow92) [inline flank-scripts/src/main/kotlin/flank/scripts/pullrequest/SetAssignees.kt:13]: Here we can use ```.appendHeaders(githubToken)``` too

Comment (piotradamczyk5) [reply flank-scripts/src/main/kotlin/flank/scripts/pullrequest/SetAssignees.kt]: fixed!

File: flank-scripts/src/main/kotlin/flank/scripts/pullrequest/GitHubIssuePropertiesCopy.kt [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import flank.scripts.github.GithubPullRequest
+import flank.scripts.github.getGitHubPullRequest
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.joinAll
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+
+object CopyProperties :
+    CliktCommand(name = "copyProperties", help = "Copy properties from referanced issue to pull request") {
```
Comment (pawelpasterz) [inline flank-scripts/src/main/kotlin/flank/scripts/pullrequest/GitHubIssuePropertiesCopy.kt:18]: Small typo
```suggestion
    CliktCommand(name = "copyProperties", help = "Copy properties from referenced issue to pull request") {
```

**Review by pawelpasterz — CHANGES_REQUESTED** [review]
I've created test issue https://github.com/Flank/flank/issues/1330 and pr https://github.com/Flank/flank/pull/1331 and ran script according to the test plan. It's amazing how fast it is really like this feature. The only difference is I am assigned in PR. Also, it could be useful to copy epics

**Comment by pawelpasterz** [comment]
Any idea why is was repeated 3 times?
Screenshot 2020-11-18 at 14 41 16

**Comment by piotradamczyk5** [comment]
> Any idea why it was repeated 3 times?
> Screenshot 2020-11-18 at 14 41 16

Zen hub API set estimates each time you run a script where Github set diff only

**Comment by piotradamczyk5** [comment]
> I've created test issue #1330 and pr #1331 and ran script according to the test plan. It's amazing how fast it is, I really like this feature. The only difference is I am assigned in PR. Also, it could be useful to copy epics

Fixed! Thanks for finding it!

**Comment by pawelpasterz** [comment]
> > I've created test issue #1330 and pr #1331 and ran script according to the test plan. It's amazing how fast it is, I really like this feature. The only difference is I am assigned in PR. Also, it could be useful to copy epics
> 
> Fixed! Thanks for finding it!
> 
> In case of copy epics, it is not so easy using Zenhub API, because it does not return epic for an issue. You could just get all epics and list issues in it, so it will not be very optimal

I see, thanks for info.

**Review by pawelpasterz — APPROVED** [review]

File: flank-scripts/src/test/kotlin/flank/scripts/GithubMockServerHandler.kt [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+import flank.scripts.utils.toJson
+import kotlinx.serialization.encodeToString
+import kotlinx.serialization.json.Json
+
+fun handleGithubMockRequest(url: String, request: Request) = when {
+    url.endsWith("/git/refs/tags/success") -> request.buildResponse("", 200)
+    url.endsWith("/releases/latest") && request.containsSuccessHeader() -> request.buildResponse(
+        GitHubRelease("v20.08.0").toJson(),
+        200
+    )
```
Comment (adamfilipow92) [inline flank-scripts/src/test/kotlin/flank/scripts/GithubMockServerHandler.kt:17]: Could we use one style of argument formatting in this file?

**Review by adamfilipow92 — APPROVED** [review]
👍

**Pull request merged**